### DISCUSSION TITLE
Fill icons

### DISCUSSION THREAD

**Discussion by mittalyashu** [lucide-icons/lucide#458] [General]
This discussion is about creating fill icons.

Here's the creation of `home` icon with fill mode.

| Preview |
| --- |
| <img width="82" alt="CleanShot 2021-12-26 at 20 46 22@2x" src="https://user-images.githubusercontent.com/29014463/147412454-980b91db-2310-4f7b-b9e6-68356709b4df.png"> |

```svg
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"> <path d="M11.3861 1.21065C11.7472 0.929784 12.2528 0.929784 12.6139 1.21065L21.6139 8.21065C21.8575 8.4001 22 8.69141 22 9V20C22 20.7957 21.6839 21.5587 21.1213 22.1213C20.5587 22.6839 19.7957 23 19 23H16C15.4477 23 15 22.5523 15 22V14C15 13.4477 14.5523 13 14 13H10C9.44772 13 9 13.4477 9 14V22C9 22.5523 8.55228 23 8 23H5C4.20435 23 3.44129 22.6839 2.87868 22.1213C2.31607 21.5587 2 20.7957 2 20V9C2 8.69141 2.14247 8.4001 2.38606 8.21065L11.3861 1.21065Z" fill="black"/> </svg>
```

_The above svg code is not optimised_

**Comment by alensiljak** [comment]
It would be great if something like this could be achieved with an attribute (i.e. `fill`). There already is the fill attribute, actually, that can be used. For example: `<Home fill="red">` but the results are not always what one would expect. :)

Related to #826 in a potential attribute that would fill an existing icon. Not sure that this would work for all, but inside a bordered container, perhaps it could invert the icon. i.e. `x circle fill`

**Comment by edwardalex** [comment]
Where can I get icons for free?

**Reply by moeenio** [reply]
in this repo

**Reply by edwardalex** [reply]
Thanks

**Comment by karsa-mistmere** [comment]
Closed #893 #407 in favour of keeping all discussions related to filled icons here.

**Comment by CarlosVergikosk** [comment]
I would like to bump this, fill icons are very useful.

**Comment by danielbayley** [comment]
We could construct certain icons in a slightly different way, and do something like—for example with [`shield-half`](https://lucide.dev/icons/shield-half)—this:
~~~ xml
<svg
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path class="lucide-fill" d="M12 22s-8-4-8-10V5l8-3" />
  <path class="lucide-hide" d="M12 22V2" />
  <path d="m12 2 8 3v7c0 6-8 10-8 10" />
</svg>
~~~
which looks exactly the same, by default:

![preview](https://github.com/lucide-icons/lucide/assets/7797479/fa842bca-06dd-4069-acee-08e0019a4ac2)

Then, optionally just have something like this in your CSS, to activate the fill version:
~~~ css
.lucide-fill { fill: currentColor }
.lucide-hide { stroke: none }
~~~
resulting in:

![preview](https://github.com/lucide-icons/lucide/assets/7797479/3916b5c9-5e8f-4d2d-b435-99c07590a8a9)

Some kind of `svgo` optimisation build step could even easily be added to strip out the classes, if desired.

What do you guys think? cc: @ericfennis @jguddas @karsa-mistmere

**Comment by danielbayley** [comment]
> Preview
> <img alt="CleanShot 2021-12-26 at 20 46 22@2x" width="82" src="https://user-images.githubusercontent.com/29014463/147412454-980b91db-2310-4f7b-b9e6-68356709b4df.png">

@mittalyashu You would want to add the `fill` as well as retain the `stroke`, to avoid it looking like shit…

**Comment by aaronmahlke** [comment]
This is great! My only problem with using lucide-icons at the moment is the lack of filled icons.
I even began playing around with filled icons recently because I thought of forking and creating a filled icon version myself. However I would love the main repo to adapt this instead.

# My Thoughts 
In my test i noticed a few things. Many icons work quite well out of the box, some need some extra attention.

Here is a preview of what I have so far:
_(none of those are optimized in any way)_
![CleanShot 2024-01-09 at 15 09 37](https://github.com/lucide-icons/lucide/assets/7060981/29564cb9-ab61-4867-9902-09722b741b97)

## Icons with open ends
Especially icons that have open ends would need some adjustment. Here is my attempt at solving some of those open end icons:
I decided on using a circle as a cutout to reveal more of the other shape.


![CleanShot 2024-01-09 at 15 29 58](https://github.com/lucide-icons/lucide/assets/7060981/9ce474cc-ab10-4e59-ad65-993aee5442c9)


## Icons with no fill possible
Some icons like arrows and similar stroke icons simply don't have a good way to have a filled version. In those cases I would leave all them the same but outline them (convert the strokes to an outline).
_Outlining leads to another issue, but more on that later..._

![CleanShot 2024-01-09 at 15 34 12](https://github.com/lucide-icons/lucide/assets/7060981/cf45e902-6de3-475e-93c1-1aac28c0d8f2)

## Scaling Icons
Scaling with stroke icons is different than scaling with filled icons.
Outline strokes (in this example the lightning) scales as well as the rest of the icon. This does not happen with the original line icons, the lightning stroke width stays the same width for all icon sizes. This would be something we would have to keep in mind. One possible solution could be to adjust for multiple sizes (in my case we are looking at 24px, 16px and 12px)

![image](https://github.com/lucide-icons/lucide/assets/7060981/f988d9af-3538-4bc4-947e-53b462b8bbe3)


---

Anyway those are my first thoughts around this topic and i would love to actively push this further.

**Reply by karsa-mistmere** [reply]
One thing I feel pretty strongly about is that filled icons should not be a full separate Lucide set, but instead a group of icons that are pretty much an integral part of the core package(s).

Firstly, this leaves plenty of room for our filled icon set to grow without the need for forcibly having to add and maintain multiple different variants of the same icon.

And more importantly: not all icons _need_ a filled variant to begin with, in fact a lot of our icons are based on lines that cannot and need not be filled.

**Reply by aaronmahlke** [reply]
Yes I definitely agree. However what do you think about consistency between using filled icons and also mixing them with stroke icons?

I am afraid that the scaling could become an issue when using both next to each other.
Let's say we use two icons in 12px size: one icon is filled, let's take my `cloud-lightning` icon from my example above and pair it with an unchanged `arrow-right` icon (also scaled down to 12px). The problem I see here is you get vastly different stroke widths for filled icons that also include some sort of stroke. However if we would have an outlined version of the stroke icons they would scale the same way as the filled icons:

![image](https://github.com/lucide-icons/lucide/assets/7060981/d9a9a9fe-6dde-403a-8a1e-73dd99e9ba26)
_Here the stroke from the lightning is **not** consistent with the stroke from the arrow_


![image](https://github.com/lucide-icons/lucide/assets/7060981/8fbf9455-a465-4a50-9736-dd400f49d50d)
_Here the stroke from the lightning is consistent with the stroke from the arrow_

I do think this could also be done programmatically. Icons that don't _need_ to be filled, could automatically be outlined for the filled version.

Another issue I see if we wouldn't be doing this, is the problem on how to address the icons in code. Let's say we wan't to only use "filled" icons in our project and therefore use the `fill` attribute to change color of icons. Now we switch the icon to be an arrow for example (which does not have a filled version) and we suddenly need to change the `stroke` instead.

**Reply by tryoxiss** [reply]
Handling the diffrent stroke widths are certinly a challange. In the seccond example, you say strokes are equal (which they likely are), but notably don't *feel* equal since they are diffrent visual weight. One way to handle it may be to use stroke for those stroke like elements, but that has its own host of problems (notably having it scale disproportianly, possibly having the lightning intercept the cloud).

As for code, you could make them proxy each other. For example if you set the fill attribute and request a stroke only version due to it being only lines, the request will return the stroke version.

EDIT: I see three additional categories of icon that need attention, beyond the ones with gaps

## Slashed/`-off` Icons

For example `message-square-off`. I think these are a pretty easy fix, as we just need to fill both sides and ensure a clean gap between the halves. This would still take some manual work, though.

## Solid-shape with stuff icons

Icons that are mostly a solid shape that can be easilly filled, but need to have bits cut out. For example `message-circle-x` would be as simple as filling the message bubble and using the existing "x" shape as a cutout.

## Special Cases

EDIT: I see in your demo you had many of these solved. I am keeping this section here as despite working well and having simple solutions, they still require additional thought and rules.

While most icons fall into either the two categories mittalyashu brought up or the additional two I brought up, a few will require special attention. For example `newspaper` is mostly just a fill-and-cut situation, but the back page should probably remain outlined. Another example would be the `book` and possibly archive icons which are the same deal. Additionally icons like `gem` may require special attention.

This is a vast minority of the icons, and most likely have simple solutions, but would still require additional discussion to agree on rules for edge-cases like this.

**Comment by zoharma** [comment]
It would be great to have these at least for icons such as [circle-check](https://lucide.dev/icons/circle-check) and [circle-x](https://lucide.dev/icons/circle-x) and the other complete circles and squares.
At the moment there's a need to target the internal `path` and `fill` to make it happen, essentially creating a subset of icons within the application.

_(update with new icon names and links)_

**Comment by mblode** [comment]
What if Lucide used Tabler's filled icons as a starting point. Tabler seems to use the same fundamental rules as Feather/Lucide. https://github.com/tabler/tabler-icons/tree/main/icons/filled

https://tabler.io/icons

I'm curious why there's never been a discussion of lucide and tabler merging the icon libraries with the substantial overlap

**Comment by jamiemlaw** [comment]
I've been having a play-around with how certain icons could be filled, including the edge-case ideas mentioned above:

https://codepen.io/jamiemlaw/pen/WNqwQpG/782fa485aa0c544b06557d4e0aac47c9

It uses the same tactic as @karsa-mistmere's https://jsfiddle.net/cgd9a3h0/1/ by filling in parts of the icon and then using a mask to invert the strokes. I really like that as a concept. However, I'm not using classes, but instead hardcoding the fills and strokes, and treating the filled icons as separate entities to their outline counterparts.

For icons like `calendar-clock` and `cloud-drizzle` I used an extra-wide stroke to carve out space for the detail on top, which feels like  quite a flexible solution.

![image](https://github.com/user-attachments/assets/2733be8f-21a5-4436-b559-060c024ecd74)

For details like the clock on the calendar or the heart on the book, I considered making these filled too, but I kinda prefer them as strokes.

**Reply by tryoxiss** [reply]
These look great, although the gem and raincloud look a bit funky in perhaps not a good way.

**Comment by hexxt-git** [comment]
how can i fill an icon with a tailwind variable?

**Reply by tryoxiss** [reply]
This is a dicussion about adding filled versions of the icons. For some icons, you could use the `fill-current` tailwind class but that will break many icons.

**Comment by aervxa** [comment]
new year, new bump

**Comment by nillpoe** [comment]
![image](https://github.com/user-attachments/assets/051e254c-c765-4405-9420-f61de0fd2dc4) (default)
![image](https://github.com/user-attachments/assets/98023916-e922-490c-b7ac-0e5fb8e346b6) (fill-current)
![image](https://github.com/user-attachments/assets/8c26c4e1-16be-4c2b-84f3-f6a3935db878) (fill-current stroke-none)
![image](https://github.com/user-attachments/assets/38a4f681-6591-4518-8e13-439f11c2c815) (expected)

**especially thumbs up/down.**

**Comment by jguddas** [comment]
![](https://github.com/user-attachments/assets/f54e442c-cde4-47d6-9ab9-d1a3c0c8707f)

**Comment by WebDev-Fish** [comment]
does anyone have any progress on this?

**Reply by jamiemlaw** [reply]
Very much still in the proof-of-concept stage; don't expect official support for filled icons any time soon. However, if you have experience working with SVGs, you can certainly take inspiration from this thread and implement them yourself in your project.

Here's an example of a how you could make a custom filled house icon in React:

![image](https://github.com/user-attachments/assets/09ce0249-ff99-4204-9261-68537a52eac3)


```js
// my-custom-filled-icons.js
export const houseFilled = [
  [
    "path",
    {
      "d": "M5 21a2 2 0 0 1-2-2v-9a2 2 0 0 1 .709-1.528l7-6a2 2 0 0 1 2.582 0l7 6A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2h-3a1 1 0 0 1-1-1v-7a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v7a1 1 0 0 1-1 1z",
      "fill": "currentColor",
      "key": "1xsl9c"
    }
  ],
];
```

```jsx
// App.jsx
import { House, Icon } from 'lucide-react';
import { houseFilled } from './my-custom-filled-icons;

const App = () => {
  return (
    {isHomePage ? <Icon iconNode={houseFilled} /> : <House />}
  );
};

export default App;
```

Or vanilla SVG

```svg
<svg
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <path d="M5 21a2 2 0 0 1-2-2v-9a2 2 0 0 1 .709-1.528l7-6a2 2 0 0 1 2.582 0l7 6A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2h-3a1 1 0 0 1-1-1v-7a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v7a1 1 0 0 1-1 1z" fill="currentColor" />
</svg>
```

**Reply by Maxx59** [reply]
@jamiemlaw How to fill the "users-round" icon?

```svg
<svg 
    xmlns="http://www.w3.org/2000/svg" 
    width="24" height="24" 
    viewBox="0 0 24 24" 
    fill="none" 
    stroke="currentColor" 
    stroke-width="2" 
    stroke-linecap="round" 
    stroke-linejoin="round" 
    class="lucide lucide-users-round-icon lucide-users-round"
>
    <path d="M18 21a8 8 0 0 0-16 0"/>
    <circle cx="10" cy="8" r="5"/>
    <path d="M22 20c0-3.37-2-6.5-4-8a5 5 0 0 0-.45-8.3"/>
</svg>
```

**Reply by jamiemlaw** [reply]
Personally, I'd cap off the bottom of the front user before filling it in.

```
<svg 
    xmlns="http://www.w3.org/2000/svg" 
    width="24" height="24" 
    viewBox="0 0 24 24" 
    fill="none" 
    stroke="currentColor" 
    stroke-width="2" 
    stroke-linecap="round" 
    stroke-linejoin="round" 
    class="lucide lucide-users-round-icon lucide-users-round"
>
    <path d="M18 21a8 8 0 0 0-16 0z" fill="currentColor"/>
    <circle cx="10" cy="8" r="5" fill="currentColor"/>
    <path d="M22 20c0-3.37-2-6.5-4-8a5 5 0 0 0-.45-8.3"/>
</svg>
```

**Reply by Maxx59** [reply]
Thank you! But is it normal that the second user in the background remains unfilled?

**Reply by Dareh123** [reply]
Greatest

**Comment by ammar08429** [comment]
some time an icon is on an surface whose color can change based on its state like active, hover etc in that case we need to fill in the icon, most of your icons already support filling, some has issues like the hand icon raised her

https://github.com/lucide-icons/lucide/issues/3403

**Reply by HeapCreep** [reply]
> some time an icon is on an surface whose color can change based on its state like active, hover etc in that case we need to fill in the icon, most of your icons already support filling, some has issues like the hand icon raised her
> 
> #3403

The same issue is happening to me now with the Tag and Tags icons: https://lucide.dev/icons/tag

Filled just shows a single dot. The icon is going over an image and I can't control what the background will be. For this reason, I want it to be filled so it stands out in more instances.

**Comment by mrgwd** [comment]
Hey everyone,
I came across this discussion while looking for a filled version of Lucide icons.
I'm a designer and developer, and I really admire the library, so I'd like to offer some help with adding filled variants to the icons, following the existing rules. I can offer about 4 hours per week. Let me know if that's something you'd be open to!

**Comment by tdzl** [comment]
Would really appreciate filled versions for on/off toggled or focused states. e.g. When a user clicks a button icon to toggle a feature on, it would be great to fill that icon in.

**Comment by StefanH-AT** [comment]
Just chiming in to say I'm also really looking forward to filled states, especially for things like `triangle-alert` which is quite hard to read on small scale and doesn't call attention to itself enough. A filled state would per perfect for that!

If there's a beta version or something I'm gonna be there as soon as possible!

**Reply by jamiemlaw** [reply]
I can't speak for the project maintainers, but none of the imlpementations proposed so far (including from me) has sparked enough joy.

The current approach would still be to write your own custom icon. As I mentioned in an earlier response, you can take inspiration from [my example Codepen](https://codepen.io/jamiemlaw/pen/WNqwQpG/782fa485aa0c544b06557d4e0aac47c9) for how some filled icons could be implemented, and then import them as custom icons (see the "see it in action" tab of the [`broom`](https://lucide.dev/icons/lab/broom) icon, for example, to see how to do that.

But, here is also a direct example of how you could implement a custom `triangle-alert-filled` icon in Vanilla JavaScript.

```js
import { createIcons } from 'lucide';

const TriangleAlertFilled = [
  [
    "defs",
    {},
    [
      [
        "mask",
        {
          "id": "lucide-triangle-alert-filled-mask"
        },
        [
          [
            "rect",
            {
              "width": 24,
              "height": 24,
              "fill": "#fff",
              "stroke": "none"
            }
          ],
          [
            "g",
            {
              "stroke": "#000"
            },
            [
              [
                "path",
                {
                  "d": "M12 9v4"
                }
              ],
              [
                "path",
                {
                  "d": "M12 17h.01"
                }
              ]
            ]
          ]
        ]
      ]
    ]
  ],
  [
    "path",
    {
      "d": "m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3",
      "fill": "currentColor",
      "mask": "url(#lucide-triangle-alert-filled-mask)"
    }
  ]
];

const icon = document.createElement('i');
icon.setAttribute('data-lucide', 'triangle-alert-filled');
document.body.appendChild(icon);

createIcons({
  icons: {
    TriangleAlertFilled,
  },
});
```

([Link to Codepen](https://codepen.io/jamiemlaw/pen/RNrMbqz?editors=1010))

**Reply by karsa-mistmere** [reply]
@jamiemlaw, the issue with this approach is using the same filled icon multiple times on a single page will result in invalid HTML (since ID attributes must be unique across the entire page). 🫤

**Reply by jamiemlaw** [reply]
@karsa-mistmere oh yeah, 100% agree. It's a passable solution to the problem but it's far from perfect.

**Comment by 1mehdifaraji** [comment]
```
import { Wallet } from "lucide-react";

<Wallet className="size-10 md:size-18 text-[#0D3380] fill-red-500" />
```
Currently using `tailwindcss` fill classname doesn't work properly on icons imported from `lucide-react`.

<img width="96" height="82" alt="Screenshot 2025-12-15 090921" src="https://github.com/user-attachments/assets/73b072d7-31a1-4c79-b01e-917a190a0dbc" />

Is there anyway to fix this issue?

**Reply by nillpoe** [reply]
there's no way to do that

**Comment by choi-me** [comment]
Any updates? REALLY needs fill icons...

**Reply by jamiemlaw** [reply]
All I can say for now is, "see previous response": if you need filled icons, the best approach right now is to take the icon you need and adapt it yourself. If the guide I posted above is lacking in any way, I can go and update it.

The main issue is that there doesn't seem to be a one-size-fits-all approach that works for all icons, so they won't become official any time soon, by my reckoning.

---

Another idea I have floating around is to mark certains paths as "accents" and others as "tints". Take [`book-heart`](https://lucide.dev/icons/book-heart), for example

![book-heart-1](https://github.com/user-attachments/assets/97bec46b-9bf1-4d12-9c50-fae84e9e77c8)

```svg
<svg
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
  class="lucide lucide-book-heart-icon lucide-book-heart"
>
  <path d="M4 19.5v-15A2.5 2.5 0 0 1 6.5 2H19a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1H6.5a1 1 0 0 1 0-5H20"/>
  <path d="M8.62 9.8A2.25 2.25 0 1 1 12 6.836a2.25 2.25 0 1 1 3.38 2.966l-2.626 2.856a.998.998 0 0 1-1.507 0z"/>
</svg>
```

Imagine adding a class of `lucide-accent` to the heart and adding an extra `lucide-tint` element that sits behind the rest of the icon

```svg
<svg
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
  class="lucide lucide-book-heart-icon lucide-book-heart"
>
  <path d="M4 19.5v-15A2.5 2.5 0 0 1 6.5 2H19a1 1 0 0 1 1 1v14h-13.5a2.5 2.5 0 0 0-2.5 2.5" class="lucide-tint" stroke="none"/>
  <path d="M4 19.5v-15A2.5 2.5 0 0 1 6.5 2H19a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1H6.5a1 1 0 0 1 0-5H20"/>
  <path d="M8.62 9.8A2.25 2.25 0 1 1 12 6.836a2.25 2.25 0 1 1 3.38 2.966l-2.626 2.856a.998.998 0 0 1-1.507 0z" class="lucide-accent"/>
</svg>
```

Along with some CSS

```css
.lucide-accent {
  stroke: oklch(58.6% 0.253 17.585);
}

.lucide-tint {
  fill: oklch(58.6% 0.253 17.585 / 0.3);
}
```

that could become

![book-heart-2](https://github.com/user-attachments/assets/2edb5c91-82da-4a19-81ca-3faeffe60c27)

And if you set the colours appropriately, you can go full fill:

```css
.lucide-accent {
  stroke: white;
}

.lucide-tint {
  fill: currentColor;
}
```

![book-heart-2](https://github.com/user-attachments/assets/1e0b6810-0c0b-4a73-a289-4f00a4bef7fc)

The advantages of this is that you can achieve both duotone and filled icons in one go, and it also sidesteps the issue of needing IDs on elements. The big disadvantage is that for filled icons, you have to explicitly set the accent colour to match the icon's background, rather than have the accent strokes be punched out and inherit the background colour automatically.
